app/src/main/kotlin/com/hippo/ehviewer/client/EhTagDatabase.kt

```diff
@@ -62,53 +62,37 @@ object EhTagDatabase : CoroutineScope {
 
     fun getTranslation(prefix: String? = NAMESPACE_PREFIX, tag: String?): String? = tagGroups[prefix]?.get(tag)?.trim()?.ifEmpty { null }
 
-    private fun rawSuggestSequence(tags: Map<String, String>, keyword: String, translate: Boolean, exactly: Boolean) = sequence {
-        if (exactly) {
-            tags[keyword]?.let {
-                yield(keyword to it.takeIf { translate })
+    context(Context)
+    fun suggestion(rawKeyword: String, expectTranslate: Boolean): Sequence<Pair<String, String?>> {
+        if (!initialized) return emptySequence()
+        val translate = expectTranslate && isTranslatable(implicit<Context>())
+        val keyword = PREFIXES.fold(rawKeyword) { kwd, pfx -> kwd.removePrefix(pfx) }
+        val prefix = rawKeyword.dropLast(keyword.length)
+        val ns = keyword.substringBefore(':')
+        val tag = keyword.drop(ns.length + 1)
+        val nsPrefix = TagNamespace.from(ns)?.prefix ?: ns
+        val tags = tagGroups[nsPrefix.takeIf { tag.isNotEmpty() && it != NAMESPACE_PREFIX }]
+        fun suggestOnce(exactly: Boolean) = let {
+            fun lookup(tags: Map<String, String>, keyword: String) = when {
+                exactly -> tags[keyword]?.let { t -> sequenceOf(keyword to t.takeIf { translate }) } ?: emptySequence()
+                translate -> tags.asSequence().filter { (tag, hint) ->
+                    tag != keyword && (tag.containsIgnoreSpace(keyword) || hint.containsIgnoreSpace(keyword))
+                }.map { (tag, hint) -> tag to hint }
+                else -> tags.keys.asSequence().filter { tag ->
+                    tag != keyword && tag.containsIgnoreSpace(keyword)
+                }.map { tag -> tag to null }
             }
-        } else {
-            if (translate) {
-                tags.forEach { (tag, hint) ->
-                    if (tag != keyword && (tag.containsIgnoreSpace(keyword) || hint.containsIgnoreSpace(keyword))) {
-                        yield(tag to hint)
-                    }
-                }
-            } else {
-                tags.keys.forEach { tag ->
-                    if (tag != keyword && tag.containsIgnoreSpace(keyword)) {
-                        yield(tag to null)
+            when {
+                tags != null -> lookup(tags, tag).map { (tag, hint) -> "$prefix$nsPrefix:$tag" to hint }
+                else -> tagGroups.asSequence().flatMap { (key, value) ->
+                    val notNsPrefix = key != NAMESPACE_PREFIX
+                    lookup(value, keyword).map { (found, hint) ->
+                        (if (notNsPrefix) "$prefix$key:$found" else "$prefix$found:") to hint
                     }
                 }
             }
```

Comment:
It's not that expensive, but the if check can be lifted outside. And it's not necessary to use `when` since there are only two branches.
```suggestion
            if (tags != null) {
                lookup(tags, tag).map { (tag, hint) -> "$prefix$nsPrefix:$tag" to hint }
            } else {
                tagGroups.asSequence().flatMap { (nsPrefix, tags) ->
                    if (nsPrefix != NAMESPACE_PREFIX) {
                        lookup(tags, keyword).map { (tag, hint) -> "$prefix$nsPrefix:$tag" to hint }
                    } else {
                        lookup(tags, keyword).map { (ns, hint) -> "$prefix$ns:" to hint }
                    }
                }
            }
```